pkg/cli/dump_test.go

```diff
@@ -490,3 +490,115 @@ ALTER TABLE orders VALIDATE CONSTRAINT fk_customer;
 		t.Fatalf("expected: %s\ngot: %s", want1, dump1)
 	}
 }
+
+func TestDumpAllTables(t *testing.T) {
```

Comment:
So postgres and thus we allow references across databases -- you can have an FK or a view in one DB refer to a table in another db, so we should be sure to include such a case here. I'm guessing that this means that above, you want to collect/accumulate all the tables for all the dbs _before_ you move on to actually dumping, and just emit all the create database statements together up front, then let the existing sort/emit/validate code run as-is, just over a larger accumulated set of tables.

Reply:
good point, thanks.

Reply:
updated code and added test case to reflect your suggestions

Reply:
nice! can you add one more test case where we have _no_ user-created databases?


I think we also want to be dumping `defaultdb` since users can create tables in that one.

Reply:
I have intentionally omitted dumping of system databases, this is why I am using following query https://github.com/cockroachdb/cockroach/pull/47387/files#diff-57fbab5a71a2e66e38932592db7fbb9bR59

```
rows, err := conn.Query(`SELECT name FROM system.namespace WHERE id > $1 AND "parentID" = 0`, []driver.Value{minUserDescID})
```

Reply:
I guess, my question what would be the expected behavior? Shall `--dump-all` list and dump all databases including system? Add explicit flag to dump system databases?

Reply:
i think system dbs in which users cannot create tables make sense to exclude, but `defaultdb` is special in that you _can_ create tables in it (i think?).

I'd say try to verify with a thought experiment / test-case: connect without a db specified (so you're using defaultdb), create a table, then dump with `--all`. I think a user could _expect_ the table they just created to be dumped. Thus I think we probably _should_ be dumping `defaultdb`. The tricky thing is that we likely can _not_ emit a CREATE for it (unless we add `IF NOT EXISTS`) because the importing cluster has its own `defaultdb`.

Reply:
> The tricky thing is that we likely can _not_ emit a CREATE for it (unless we add `IF NOT EXISTS`) because the importing cluster has its own `defaultdb`.

Now, when I think of that comments, it makes me wonder why `IF NOT EXISTS` is not appended now? I mean, for obvious reason `defaultdb` will exists (while honestly do not think it should be any different from normal databases), but there might be a collision with other names intentionally or not (normal names usually have very low entropy).

Reply:
> i think system dbs in which users cannot create tables make sense to exclude, but `defaultdb` is special in that you _can_ create tables in it (i think?).

totally make sense, but is there a way to distinguish dbs where user can create tables versus those where he cannot?

Reply:
also it makes me wonder how the query

```
SELECT name FROM system.namespace WHERE id > $1 AND "parentID" = 0
```

didn't hit `defaultdb`

Reply:
another thought, given we dump `defaultdb` would you expect to have 

```
CREATE DATABASE defaultdb;
```

in the resulted outcome? or since this is a default and potentially always existing database it should be 

```
USE defaultdb;
```

instead?

Reply:
> I'd say try to verify with a thought experiment / test-case: connect without a db specified (so you're using defaultdb), create a table, then dump with `--all`. I think a user could _expect_ the table they just created to be dumped.

I've came up with updated code + added a unit-test to cover the scenario with `defaultdb`.